# Introduction to Counting, Permutations and Combinations

![](https://i.imgur.com/B2lB620.jpg)

Solving problems in statistics and probability often involves counting events satisfying a criteria without listing them out. In this tutorial, we'll learn some basic counting techniques that can be combined in interesting ways. This tutorial covers the following topics:

- Multiplication principle
- Factorial rule
- Permutations rule
- Combinations rule
- Computing probability using counting rules

## Multiplication Principle of Counting

> **QUESTION**: Your neighborhood Mexican restaurant "Tacos El Vegano" offer a "build your own burrito" option and claims that you create over 250 unique varieties by making the following choices:
>
> 
> 
> 1. Choice of tortilla: white flour or whole wheat flour
> 2. Choice of filling: potatoes, jackfruit or mushroom
> 3. Choice of beans: red beans, black beans, kidney beans or no beans
> 4. Choice of salad: lettuce, rocket or no salad
> 5. Choice of grilled vegetables: yes or no
> 6. Choice of guacamole: yes or no
>
> <img src="https://i.imgur.com/1vx5O8w.png" width="360" style="border-radius:4px">
>
> All the choices are made independently i.e. the choice of beans has no bearing on whether or not you can add guacamole and vice versa. Can you verify or refute this claim?
>




One way to check the validity of the claim is to systematically list all possible varieties:

1. White flour, Potatoes, Red beans, Lettuce, Vegetables, Gaucamole
2. White flour, Potatoes, Red beans, Lettuce, Vegetables, No Gaucamole
3. White flour, Potatoes, Red beans, Lettuce, No Vegetables, Gaucamole
4. White flour, Potatoes, Red beans, Lettuce, No Vegetables, No Gaucamole
5. White flour, Potatoes, Red beans, Rocket, Vegetables, Gaucamole
6. ...


This is a long, tedious and error prone process. Fortunately, there's an easier way:

![](https://i.imgur.com/nf3OD3L.png)


* There are 2 choices for the tortilla (white flour or whole wheat flour). 
* Once a choice of tortilla is made, then the filling can be chosen in 3 ways (potatoes, jack fruit or mushroom) for each of the two choices. So, if we just consider the first two choices, we can create 3 + 3 = 2 * 3 = 6 total varieties. 
* Next, for each of these 6 combinations, we have 4 choices of beans (red beans, black beans, kidney beans or no beans). Thus the total number of varieties considering the first three choices are is 4 + 4 + 4 + 4 + 4 + 4 = 6 * 4 = 24.
* and so on...

If you keep going, you will realize that the total number of varieties is simply the product of the number of options for each choice. We can compute this using Python:

In [1]:
n_tortilla = 2
n_filling = 3
n_beans = 4
n_salad = 3
n_vegetables = 2
n_guacamole = 2

In [2]:
n_variations = n_tortilla * n_filling * n_beans * n_salad * n_vegetables * n_guacamole

In [3]:
print('Tacos El Vegano offers {} variations of burritos.'.format(n_variations))

Tacos El Vegano offers 288 variations of burritos.


Turns out the restaurant's claim is valid after all! We can generalize the above result as follows:

**Multiplication Principle**:  If we have $k$ independent choices to make, and the number of options for each choice are $x_1, x_2, x_3, \ldots, x_k$, then we can make all of the choices in $x_1 \times x_2 \times x_3 \times \ldots \times x_k$ ways.

A special case of the multiplication principle is when the number of options is the same for each choice. This happens when we make the same choice many times.  

**Repeated Choice Rule**: If we have $k$ independent choices to make, and the number of options for each choice is $n$, then we can make all the choices in $n^k$ ways.



> **EXERCISE**: If you flip 10 coins, how many possible outcomes can you get?

In [5]:
k = 10
n=2
outcome = n**k
outcome

1024

> **EXERCISE**: If you roll 6 fair dice, how many possible outcomes can you get?

In [6]:
k = 6
n = 6
outcome = n**k
outcome

46656

In [10]:
k = 6
n = 36
outcome = n//k
outcome

6

> **EXERCISE**: If you flip a coin, roll a die and pick a random card from a well-shuffled deck, how many possible outcomes can you get?

In [73]:
k = 6
n = 625
t = []
u = []
for i in range(1,100):
    outcome = n//i
    t.append(outcome)
    u.append(i)
    if i == outcome:
        print(i)

25


In [74]:
r=(t,u)

Let's save our work before continuing.

In [75]:
import jovian

In [76]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "rakesh-rajagopalachary/counting-permutations-and-combinations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/rakesh-rajagopalachary/counting-permutations-and-combinations


'https://jovian.ai/rakesh-rajagopalachary/counting-permutations-and-combinations'

## Factorials

> **QUESTION**: In how many ways can you seat 5 people on chairs numbered from 1 to 5?
> 
> <img src="https://i.imgur.com/eLo4HFD.png" width="420">


One way to determine the result is to list all possible arrangements systematically:

1. A, B, C, D, E
2. A, B, C, E, D
3. A, B, D, C, E
4. A, B, D, E, C
5. ...

This is a long, tedious and error prone process. However, we can apply the multiplication rule using the following insight:

![](https://i.imgur.com/sbCygLM.png)

1. We have 5 ways of choosing the person for seat number 1.
2. Once a person is seated at seat no. 1, we have 4 ways of choosing the person for seat number 2 (because the person sitting on seat number 1 can't sit on seat number 2). Note that we have exactly 4 choices for the seat irrespective of which person we choose for seat 1.
3. Once a person is seated at seat no. 2, we have 3 choices for choosing the person for seat number 3.
4. Once a person is seated at seat no. 3, we have to choices for choosing the person at seat number 2.
5. Once a person is seated at seat no. 3, we have just one choice to fill the remaining seat.


Thus, by applying the multiplication rule, it follows that the total number of ways to seat five persons on five chairs is $5 \times 4 \times 3 \times 2 \times 1 = 120$. 

The product of numbers from $1$ to $5$ is also known as the factorial of $5$ and is represented using by the notation $5!$. In general $n!$ denotes the product of numbers from $1$ to $n$. $0!$ is defined as $1$ (can you guess why?). With this notation, we can now generalize the above result as follows:

**Factorial Rule**: The number of ways of arranging $n$ items in list is $n!$ i.e. $n \times (n-1) \times (n-2) \times \ldots \times 2 \times 1 $.

Let's define a function factorial which can help us compute this number easily.

In [77]:
def factorial(n):
    result = 1
    for i in range(1, n+1):
        result = result * i
    return result

In [78]:
factorial(5)

120

> **EXERCISE**: If you pick out all the spades from a deck of cards, in how many ways can you arrange them in your hand?
> 
> ![](https://i.imgur.com/XVscmAe.png)
>
> *Hint*: There are 13 spades.

In [79]:
factorial(13)

6227020800

In [80]:
k = 13
n = 52
outcome = n**k
outcome

20325604337285010030592

> **EXERCISE**: If you roll 6 dice together, in how many ways can you get 6 different numbers in the result e.g.  3,4,5,6,1,2.

In [81]:
factorial(6)

720

Let's save our work before continuing.

In [82]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Updating notebook "rakesh-rajagopalachary/counting-permutations-and-combinations" on https://jovian.ai/
[jovian] Committed successfully! https://jovian.ai/rakesh-rajagopalachary/counting-permutations-and-combinations


'https://jovian.ai/rakesh-rajagopalachary/counting-permutations-and-combinations'

## Permutations

> **QUESTION**: In how many ways can you seat 6 people on 4 chairs numbered from 1 to 4?
>
> <img src="https://i.imgur.com/xVmrrBj.png" width="360">


Let's apply the multiplication rule once again:

1. We have 6 ways of choosing the person for seat number 1.
2. Once a person is sitting at seat number 1, we have 5 ways of choosing a person for seat number 2.
3. Once a person is sitting at seat number 2, we have 4 ways of choosing a person for seat number 3.
4. Once a person is sitting at seat number 3, we have 3 ways of choosing a person for seat number 1.

And we're done! Applying the multiplication rule, it follows that the number of ways of seating 6 people on 4 chairs is $6 \times 5 \times 4 \times 3 = 360$.


However, we can also write the expression has follows:

$$6 \times 5 \times 4 \times 3 = \frac {6 \times 5 \times 4 \times 3 \times 2 \times 1} {2 \times 1} = \frac{6!}{2!} = \frac{6!}{(6-4)!}$$


We can generalize the above result as follows:

**Permutation rule**: The number of ways of arranging $k$ items in a list out of $n$ items is $n!/(n-k)!$. This number is also denoted as ${}^nP_k$.

$${}^nP_k = \frac{n!}{(n-k)!} = n \times (n-1) \times \ldots \times (n-k+1)$$

Each arrangement of `k` out of `n` items is called a permutation. Let's write a helper function to count the number of permutations of `k` out of `n` items.


In [83]:
def permutations(n, k):
    return factorial(n) // factorial(n-k)

> **EXERCISE**: You have 25 books but your bookshelf contains space for just 12 books. In how many ways can you arrange books on your bookshelf?

In [86]:
permutations(25, 12)

2490952020480000

> **EXERCISE**: How many unique four letter words can you create using the English alphabet (A to Z)?

> **EXERCISE**: How many unique four letter words can you create using the English alphabet (A to Z), if you aren't allowed to use a letter more than once?

In [87]:
permutations(26, 4)

358800



> **EXERCISE**: How many 3 digit numbers greater than 300 can you create using the digit 1, 2, 3, 5, 7 and 8? You are not allowed to use a digit more than once. How many of these numbers are odd?

> **EXERCISE**: In how many ways can you arrange 12 people in a circle?
> 
> <img src="https://i.imgur.com/J738ur4.png" width="360">

Let's save our work before continuing.

In [2]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/aakashns/counting-permutations-and-combinations


'https://jovian.ai/aakashns/counting-permutations-and-combinations'

## Combinations

> **QUESTION**: In how many ways can you choose 4 people out of a group of 6 people.

Let's try to solve this problem step by step:

* We already know from the permutation rule that we can arrange $4$ out of $6$ people in a list in ${}^6P_4$ ways. 
* However, we're no longer concerned with arrangement i.e. if we name the people A, B, C, D, E and F, then the arrangements ABCD, ACDB, ADCB all denote the same group. In fact, once $4$ people out of the $6$ have been chosen, we know from the factorial rule that there are $4!$ possible arrangements. 
* Thus, every $4!$ arrangements of any $4$ people denotes the same group. So, we can divide the total number of arrangements by $4!$ to determine the number of ways of choosing 4 people out of 6.


$$\textrm{No. of ways of choosing 4 people out of 6} = \frac{\textrm{No. of ways of arranging 4 people out of 6 in a list}}{\textrm{No. of ways of arranging 4 people in a list}} = \frac{{}^6P_4}{4!}$$

This number is denoted by ${}^6C_4$. Simplifying, we get:

$${}^6C_4 = \frac{{}^6P_4}{4!} = \frac{6!}{(6-4)!4!}$$


The above result can be generalized as follows:

**Combination Rule**: The number of ways of choosing $k$ items out of $n$ items is $\frac{n!}{(n-k)!k!}$. It is denoted by ${}^nC_k$.

$${}^nC_k = \frac{{}^nP_k}{k!} = \frac{n!}{(n-k)!k!} = \frac{n \times (n-1) \times \ldots \times (n-k+1)}{k \times (k-1) \times \ldots \times 1}$$

Let's write a function to count the number of ways of choosing `k` items out of `n`.

In [88]:
def combinations(n, k):
    return permutations(n, k) // factorial(k)

In [89]:
combinations(6, 4)

15

Thus, there are 15 ways of choosing 4 people out of 6. Can you verify this by listing out all possible groups?

> **EXERCISE**: In how many ways can 22 soccer players be divided into two teams of 11 players each?

> **EXERCISE**: In how many ways can you get exactly 3 heads when you toss 10 fair coins?


> **EXERCISE**: Suppose you pick 4 letters from the English alphabet (A to Z). How many words can you create using the 4 letters (you are allowed to use a letter more than once in a word)?

Let's save our work before continuing.

In [20]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "aakashns/counting-permutations-and-combinations" on https://jovian.ai/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/aakashns/counting-permutations-and-combinations


'https://jovian.ai/aakashns/counting-permutations-and-combinations'

## Probabilities

The rules of counting can be applied to compute probabilities events involving repeated experiments (or multiple different kinds of experiments). See this tutorial for a practical introduction to probability: https://jovian.ai/aakashns/introduction-to-probability .

> **QUESTION**: If you toss 10 fair coins, what is the probability of getting exactly 3 heads?

Since all possible outcomes are equally likely, it follows that

$$P(\textrm{Getting 3 heads}) = \frac{\textrm{No. of outcomes with exactly 3 heads}}{\textrm{Total no. of outcomes}}$$

$$\textrm{Total no. of outcomes} = {2 \times 2 \times ... \times 2}  \textrm{(10 times)} = 2^{10}$$

$$\textrm{No. of outcomes with exactly 3 heads} = {}^{10}C_3 = \frac{10!}{(10-3)!3!}$$

We can use the functions defined earlier to compute this number.

In [21]:
total_outcomes = 2**10
matching_outcomes = combinations(10, 3)
p_3_heads = matching_outcomes / total_outcomes

In [24]:
print('The probability of getting exactly 3 heads when 10 fair coins are tossed is {:.2f} .'.format(p_3_heads))

The probability of getting exactly 3 heads when 10 fair coins are tossed is 0.12 .


> **QUESTION**: If you pick 5 random cards from a well shuffled deck, what is the probability of getting a flush (all cards of the same suit)?

We can determine the total no. of outcomes by counting the number of ways of picking 5 cards out of a deck of 52 cards.

$$\textrm{Total no. of outcomes} = {}^{52}C_5$$

To count the number of outcomes, we can first determine the suit in 4 ways, and then pick of 5 out of 13 cards in the suit.

$$\textrm{Matching outcomes} = 4 \times {}^{13}C_5$$

We can now compute the probability that all 5 cards belong to the same suit.

In [25]:
total_outcomes = combinations(52, 5)
matching_outcomes = 4 * combinations(13, 5)
p_flush = matching_outcomes / total_outcomes

In [31]:
print('If 5 cards are picked from well shuffled deck, the probability of getting a flush is {:.3f} .'.format(p_flush))

If 5 cards are picked from well shuffled deck, the probability of getting a flush is 0.002 .


> **EXERCISE**: If you toss 12 fair coins what is the probablity of getting 7 or more heads?
> 
> *Hint*: $P(\textrm{7 or more heads}) = P(\textrm{7 heads}) + P(\textrm{8 heads}) + P(\textrm{9 heads}) + \ldots$

> **EXERCISE**: If you roll 6 fair dice together, what is the probablity of getting six different numbers in the result?

> **EXERCISE**: If you flip a coin, roll a die and pick a random card from a well-shuffled deck, what is the probability of getting a tail, an even number and a spade?

> **EXERCISE**: If you roll 3 fair dice, what is the probability of getting a sum lower than 6?

Let's save our work before continuing.

In [2]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/aakashns/counting-permutations-and-combinations


'https://jovian.ai/aakashns/counting-permutations-and-combinations'

Let's save our work before continuing.

In [2]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ai/aakashns/counting-permutations-and-combinations


'https://jovian.ai/aakashns/counting-permutations-and-combinations'

## Summary and Further Reading

Here are the topics we've covered in this notebook:


1. **Multiplication Principle**:  If we have $k$ independent choices to make, and the number of options for each choice are $x_1, x_2, x_3, \ldots, x_k$, then we can make all of the choices in $x_1 \times x_2 \times x_3 \times \ldots \times x_k$ ways.


2. **Repeated choice rule**: A special case of the multiplication principle is when the number of options is the same for each choice. If we have $k$ independent choices to make, and the number of options for each choice is $n$, then we can make all the choices in $n^k$ ways.


2. **Factorial Rule**: The number of ways of arranging $n$ items in list is $n!$ i.e. $n \times (n-1) \times (n-2) \times \ldots \times 2 \times 1 $.


3. **Permutation rule**: The number of ways of arranging $k$ items in a list out of $n$ items is $n!/(n-k)!$. This number is also denoted as ${}^nP_k$.

$${}^nP_k = \frac{n!}{(n-k)!} = n \times (n-1) \times \ldots \times (n-k+1)$$


4. **Combination Rule**: The number of ways of choosing $k$ items out of $n$ items is $\frac{n!}{(n-k)!k!}$. It is denoted by ${}^nC_k$.

$${}^nC_k = \frac{{}^nP_k}{k!} = \frac{n!}{(n-k)!k!} = \frac{n \times (n-1) \times \ldots \times (n-k+1)}{k \times (k-1) \times \ldots \times 1}$$


5. **Probability**: The rules of counting can be applied to compute probabilities events involving repeated experiments (or multiple different kinds of experiments).


Resources to learn more:

* Khan Academy: https://www.khanacademy.org/math/statistics-probability/counting-permutations-and-combinations
* Permutations and Combinations problems: https://www.analyzemath.com/statistics/permutations_combinations.html
* Better Explained: https://betterexplained.com/articles/easy-permutations-and-combinations/

## Questions for Revision
1.	Explain multiplication rule with an example.
2.	What is the special case of multiplication rule?
3.	In how many ways can we makes choices when the number of independent choices is 'q' and number of options for each choice is 'r'?
4.	Where do we generally come across the repeated choice rule?
5.	What is the factorial rule?
6.	There are 10 benches in a classroom placed one after another. In how many ways can the 10 students be seated on the benches for an examination?
7.	What are some examples of permutations and combinations that we come across on a daily basis?
8.	How can you identify that a problem refers to either permutation or combination?
9.	How is permutation different from combination?
10.	Is there any relationship between permutation and combination? If so, what is it?
11.	What are the rules of counting that can be applied to compute probabilities? 
12.	Give an example for permutation when repetition is allowed and when repetition is not allowed.
13.	Give an example for combination when repetition is allowed and when repetition is not allowed.
14.	You have 7 ice-cream flavours say- chocolate, tender coconut, mango, pistachio, vanilla, butterscotch, and blackcurrant. You choose any four flavours in each cup. How many variations can there be? 
15.	In how many ways can the word JOVIAN be arranged?

## Solutions for Exercises

### Multiplication Principle of Counting

If we have $k$ independent choices to make, and the number of options for each choice are $x_1, x_2, x_3, \ldots, x_k$, then we can make all of the choices in $x_1 \times x_2 \times x_3 \times \ldots \times x_k$ ways.


**Repeated Choice Rule**: If we have $k$ independent choices to make, and the number of options for each choice is $n$, then we can make all the choices in $n^k$ ways.

> **EXERCISE**: If you flip 10 coins, how many possible outcomes can you get?

- We have two options for each choice i.e Heads or Tails
- And we have a total of 'k' independent Choices = 10
- Option for each choice = n = 2

In [2]:
k = 10
n = 2
n_outcomes = n**k
n_outcomes 

1024

> **EXERCISE**: If you roll 6 fair dice, how many possible outcomes can you get?

- Each die can return a number between 1 and 6 at each roll, therefore number of options at each independent choice is n = 6

- the number of independant choices(k) = 6

- There fore total outcomes can be: 

In [3]:
k = 6
n = 6
n**k

46656

> **EXERCISE**: If you flip a coin, roll a die and pick a random card from a well-shuffled deck, how many possible outcomes can you get?

- Sample outcomes could be: (H,6,Spade-7), (H,1, Heart-11) and so on, 

So, 
- *Flipping a coin:* gives us 2 options on one throw, $=x_1$
- *Rolling a die:* 6 options on one throw, $=x_2$ and
- *Picking a card:* 52 options in one throw,$=x_3$

- So possible number of outcomes will be $= x_1 \times x_2 \times x_3$

In [4]:
2*6*52

624

### Factorials

> **QUESTION:** In how many ways can you seat 5 people on chairs numbered from 1 to 5?

In [3]:
def factorial(n):
    result = 1
    for i in range(1, n+1):
        result = result * i
    return result

In [4]:
factorial(5)

120

> **EXERCISE**: If you pick out all the spades from a deck of cards, in how many ways can you arrange them in your hand?
> 
> ![](https://i.imgur.com/XVscmAe.png)
>


- Total number of spades: 13
- Number of ways to arrange 'n' things in a sequence: factorial(n)

In [5]:
factorial(13)

6227020800

> **EXERCISE**: If you roll 6 dice together, in how many ways can you get 6 different numbers in the result e.g.  3,4,5,6,1,2.

- Total combinations when we roll 6 die's together $= 6^6 = 46656$ 
- Getting 6 different numbers at each throw = $6\times5\times4\times3\times2\times1 $ i.e $6!$

In [9]:
factorial(6)

720

### Permutations

In [4]:
def permutations(n, k):
    return factorial(n) // factorial(n-k)

> **EXERCISE**: You have 25 books but your bookshelf contains space for just 12 books. In how many ways can you arrange books on your bookshelf?

-  $25$ Total Books and Space for $12$  
- Using Permutations, ${}^nP_k = \frac{n!}{(n-k)!}$
- $n = 25, k = 12$

In [5]:
permutations(25,12)

2490952020480000

> **EXERCISE**: How many unique four letter words can you create using the English alphabet (A to Z)?

- Note that: while we are creating a word we can use repeated alphabets.
- We have 26 Total alphabets 
- We have to create a word of 4 characters, meaning we have 26 options for each individual choice of the 4 letters. 

- So, using *Repeated Choice Rule*, we have $n^k = {26}^4$ options to choose from.  

In [6]:
k = 4
n = 26

n**k

456976

> **EXERCISE**: How many unique four letter words can you create using the English alphabet (A to Z), if you aren't allowed to use a letter more than once?

- Since we cannot use a letter more than once, we are left with a permutation of arranging 26 alphabets into 4 letters.
- ${}^nP_k = ^{26}P_4$

In [7]:
permutations(26,4)

358800

> **EXERCISE**: How many 3 digit numbers greater than 300 can you create using the digit 1, 2, 3, 5, 7 and 8? You are not allowed to use a digit more than once. How many of these numbers are odd?

**Solution:**
- Let's consider 2 Sets depending on first digit for our three digit number: (We can only choose from 3,5,7,8 to make our number > 300)
    - (odd) Set 1: Where first digit can be: 3 _ _,5 _ _,7 _ _ 
    - (even) Set 2: Where first digit can be: 8 _ _  
    
- Set 1: 
    - First digit: 3 options to choose from as mentioned above.
    - Last digit: {1,3,5,7}i.e 4 options to make our number odd, but one of 3,5,7 has already been used so 3 options to choose from.  
    - Middle digit: 6 total options but 2 have been used already so 4 options to choose from. 
    - So for Set-1, there are $3\times3\times4 $ possible ways.
- Set 2: 
    - First digit: 1 option as mentioned above.
    - Last digit: {1,3,5,7}i.e 4 options to make our number odd, so 4 options to choose from.  
    - Middle digit: 6 total options but 2 have been used already so 4 options to choose from. 
    - So for Set-1, there are $1\times4\times4 $ possible ways.

In [6]:
set_1 = 3*3*4
set_2 = 1*4*4

total_nums = set_1 + set_2
total_nums

52

> **EXERCISE**: In how many ways can you arrange 12 people in a circle?
> 
> <img src="https://i.imgur.com/J738ur4.png" width="360">

**Solution:**

 **Circular Permutations:** Unlike linear arrangements, circular arrangements don't have a start and an end. For linear arrangements, total number of ways to arrange n objects is n! But for circular permutations total number of ways to arrange n objects is (n-1)!

Read more about [Circular Permutations](https://doubleroot.in/lessons/permutations-combinations/circular-permutations/)

- 12 people, therefore (12-1)! = 11!

In [7]:
factorial(11)

39916800

### Combinations

In [9]:
def combinations(n, k):
    return permutations(n, k) // factorial(k)

> **EXERCISE**: In how many ways can 22 soccer players be divided into two teams of 11 players each?

- After we choose 11 players, other 11 are automatically chosen. 

In [11]:
combinations(22,11)

705432

> **EXERCISE**: In how many ways can you get exactly 3 heads when you toss 10 fair coins?


In [12]:
combinations(10,3)

120

> **EXERCISE**: Suppose you pick 4 letters from the English alphabet (A to Z). How many words can you create using the 4 letters (you are allowed to use a letter more than once in a word)?

- For Picking 4 alphabets we have $^{26}C_4$ ways.
- Since letter can be repeated 
- Using Multiplication rule: k = 4, x = 4


In [23]:
combinations(26,4) * (4**4)

3827200

### Probabilities

> **EXERCISE**: If you toss 12 fair coins what is the probablity of getting 7 or more heads?
> 
> *Hint*: $P(\textrm{7 or more heads}) = P(\textrm{7 heads}) + P(\textrm{8 heads}) + P(\textrm{9 heads}) + \ldots$

- $P(\textrm{7 heads}) = ^7C_3 /$ Total_outcomes 
- $P(\textrm{8 heads}) = ^8C_3 / $Total_outcomes 
- $P(\textrm{9 heads}) = ^9C_3 / $Total_outcomes 
- $P(\textrm{10 heads}) =^{10}C_3 / $Total_outcomes 
- $P(\textrm{11 heads}) =^{11}C_3 / $Total_outcomes 
- $P(\textrm{12 heads}) =^{12}C_3 / $Total_outcomes 

In [16]:
seven_h = combinations(12,7)
eight_h = combinations(12,8)
nine_h = combinations(12,9)
ten_h = combinations(12,10)
eleven_h = combinations(12,11)
twelve_h = combinations(12,12)

In [15]:
Total_outcomes = 2**12
Total_outcomes

4096

In [17]:
(seven_h + eight_h +nine_h + ten_h + eleven_h + twelve_h) / Total_outcomes

0.38720703125

> **EXERCISE**: If you roll 6 fair dice together, what is the probablity of getting six different numbers in the result?

- Total Outcomes = $6^6$ 
- 6 options for first dice, 5 for the second and so on...
- Therefore, 6! shall be our answer!

In [18]:
factorial(6)

720

In [25]:
probability = factorial(6) / (6**6)
probability

0.015432098765432098

> **EXERCISE**: If you flip a coin, roll a die and pick a random card from a well-shuffled deck, what is the probability of getting a tail, an even number and a spade?


- A - event of flipping a coin and getting a tail - $P(A) =  1\div2$
- B - event of rolling a die to get an even num - $P(B) = 3\div6$
- C - event, picking a random card and getting a spade - $P(C) = 13\div52$
- Since al three are independent events, $P(A)\times P(B) \times P(C) $ 

In [19]:
p_a = 1/2
p_b = 1/2
p_c = 13/52

p_a * p_b * p_c

0.0625

> **EXERCISE**: If you roll 3 fair dice, what is the probability of getting a sum lower than 6?

- We can create a sample space for getting the sum lower than 6.
- (1,1,1) (1,1,2) (1,2,1) (1,1,3) (1,3,1) (2,1,1) (2,2,1) (2,1,2) (3,1,1) (1,3,1) = 10 in Total
- Total_outcomes = $6^3$

In [21]:
Total_probability = 10/(6**3)
Total_probability

0.046296296296296294